consensus/src/liveness/proposal_generator.rs

```diff
@@ -305,6 +338,49 @@ impl ProposalGenerator {
         ))
     }
 
+    async fn calculate_max_block_sizes(&mut self, voting_power_ratio: f64, timestamp: Duration) -> (u64, u64) {
+        let mut values_max_block_txns = vec![self.max_block_txns];
+        let mut values_max_block_bytes = vec![self.max_block_bytes];
+
+        let chain_health_backoff = self
+            .chain_health_backoff_config
+            .get_backoff(voting_power_ratio);
+        if let Some(value) = chain_health_backoff {
+            values_max_block_txns.push(value.max_sending_block_txns_override);
+            values_max_block_bytes.push(value.max_sending_block_bytes_override);
+            CHAIN_HEALTH_BACKOFF_TRIGGERED.observe(1.0);
+        } else {
+            CHAIN_HEALTH_BACKOFF_TRIGGERED.observe(0.0);
+        }
+
+        let pipeline_backpressure = self
+            .pipeline_backpressure_config
+            .get_backoff(self.block_store.pipeline_pending_latency(timestamp));
+        if let Some(value) = pipeline_backpressure {
+            values_max_block_txns.push(value.max_sending_block_txns_override);
+            values_max_block_bytes.push(value.max_sending_block_bytes_override);
+            PIPELINE_BACKPRESSURE_ON_PROPOSAL_TRIGGERED.observe(1.0);
+        } else {
+            PIPELINE_BACKPRESSURE_ON_PROPOSAL_TRIGGERED.observe(0.0);
+        };
+
+        let max_block_txns = values_max_block_txns.into_iter().min().unwrap();
+        let max_block_bytes = values_max_block_bytes.into_iter().min().unwrap();
+
+        if pipeline_backpressure.is_some() || chain_health_backoff.is_some() {
+            warn!(
+                "Generating proposal: reducing limits to {} txns and {} bytes, due to pipeline_backpressure: {}, chain health backoff: {}. Delaying sending proposal by {}ms",
+                max_block_txns,
+                max_block_bytes,
+                pipeline_backpressure.is_some(),
+                chain_health_backoff.is_some(),
+                self.backpressure_proposal_delay.as_millis(),
+            );
+            tokio::time::sleep(self.backpressure_proposal_delay).await;
```

Comment:
how does this work with the wait inside pull_payload?

Reply:
wanted to leave that for later, but yeah, this should reduce the wait there, and with that we can increase this delay up to 300ms (which is the delay there), which should help quite a bit I think